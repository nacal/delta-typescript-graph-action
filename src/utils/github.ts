import * as core from '@actions/core';
import * as github from '@actions/github';
import { log } from './log';
import { execSync } from 'child_process';

export default class GitHub {
  #octokit: ReturnType<typeof github.getOctokit>;

  constructor() {
    this.#octokit = github.getOctokit(core.getInput('access-token'));
  }

  public async getTSFiles() {
    const compareResult =
      await this.#octokit.rest.repos.compareCommitsWithBasehead({
        owner: github.context.repo.owner,
        repo: github.context.repo.repo,
        basehead: `${github.context.payload.pull_request?.base.sha}...${github.context.payload.pull_request?.head.sha}`,
      });
    const files = compareResult.data.files
      ?.filter(file =>
        /\.ts$|\.tsx$|\.vue$|\.astro$|\.svelte$/.test(file.filename),
      )
      .map(file => ({
        filename: file.filename,
        status: file.status,
        previous_filename: file.previous_filename,
      }));
    log(files);

    // typescript-graph では、以下の分類で処理する。
    return {
      created:
        files?.filter(
          file => file.status === 'added' || file.status === 'copied',
        ) ?? [],
      deleted: files?.filter(file => file.status === 'removed') ?? [],
      modified:
        files?.filter(
          file => file.status === 'modified' || file.status === 'changed',
        ) ?? [],
      renamed: files?.filter(file => file.status === 'renamed') ?? [],
      unchanged: files?.filter(file => file.status === 'unchanged') ?? [],
    };
  }

  public getBaseSha() {
    return github.context.payload.pull_request?.base.sha;
  }

  public getHeadSha() {
    return github.context.payload.pull_request?.head.sha;
  }

  /**
   * コメントのタイトルは同一コメントを探す際にも使用する
   */
  public getCommentTitle() {
    return `## Delta TypeScript Graph<!--${github.context.workflow}-->`;
  }

  /**
   * PRにコメントする。
   *
   * 同一PR内では同一コメントを上書きし続ける。
   */
  public async commentToPR(body: string) {
    const owner = github.context.repo.owner;
    const repo = github.context.repo.repo;
    const issue_number = github.context.payload.number;
    github.context.workflow;
    // 1. 既存のコメントを取得する
    const comments = await this.#octokit.rest.issues.listComments({
      owner,
      repo,
      issue_number,
    });

    // 2. 既存のコメントがあれば、そのコメントのIDを取得する
    const existingComment = comments.data.find(
      comment => comment.body?.trim().startsWith(this.getCommentTitle()),
    );

    if (existingComment) {
      // 3. 既存のコメントがあれば、そのコメントを更新する
      await this.#octokit.rest.issues.updateComment({
        owner,
        repo,
        comment_id: existingComment.id,
        body,
      });
    } else {
      // 4. 既存のコメントがなければ、新規にコメントを作成する
      await this.#octokit.rest.issues.createComment({
        owner,
        repo,
        issue_number,
        body,
      });
    }
  }

  /**
   * PRのコメントを削除する
   */
  public async deleteComment() {
    const owner = github.context.repo.owner;
    const repo = github.context.repo.repo;
    const issue_number = github.context.payload.number;
    // 1. 既存のコメントを取得する
    const comments = await this.#octokit.rest.issues.listComments({
      owner,
      repo,
      issue_number,
    });

    // 2. 既存のコメントがあれば、そのコメントのIDを取得する
    const existingComment = comments.data.find(
      comment => comment.body?.trim().startsWith(this.getCommentTitle()),
    );

    if (existingComment) {
      // 3. 既存のコメントがあれば、そのコメントを削除する
      await this.#octokit.rest.issues.deleteComment({
        owner,
        repo,
        comment_id: existingComment.id,
      });
    }
  }

  public cloneRepo() {
    const repo = github.context.repo;
    // リポジトリのURLを取得
    const repoUrl = `https://github.com/${repo.owner}/${repo.repo}.git`;
    // リポジトリをチェックアウト
    execSync(`git clone ${repoUrl}`);
    // result としてリポジトリ名を返す
    return { repoDir: repo.repo };
  }
}
