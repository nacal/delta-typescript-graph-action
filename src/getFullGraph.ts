import { createGraph } from '@ysk8hori/typescript-graph/dist/src/graph/createGraph';
import { execSync } from 'child_process';
import { log } from './utils/log';
import { getTsconfigRoot, getTsconfigPath } from './utils/config';
import path from 'path';
import GitHub from './utils/github';

/**
 * TypeScript Graph の createGraph を使い head と base の Graph を生成する
 *
 * 内部的に git fetch と git checkout を実行するので、テストで実行する際には execSync を mock すること。
 *
 * また、処理に時間がかかるため Promise を返す。
 */
export default function getFullGraph() {
  const github = new GitHub();
  // head の Graph を生成するために head に checkout する
  execSync(`git fetch origin ${github.getHeadSha()}`);
  execSync(`git checkout ${github.getHeadSha()}`);
  // head の Graph を生成
  const { graph: fullHeadGraph, meta } = createGraph({
    tsconfig: getTsconfigPath(),
    dir: path.resolve(getTsconfigRoot()),
  });
  log('fullHeadGraph.nodes.length:', fullHeadGraph.nodes.length);
  log('fullHeadGraph.relations.length:', fullHeadGraph.relations.length);
  // base の Graph を生成するために base に checkout する
  execSync(`git fetch origin ${github.getBaseSha()}`);
  execSync(`git checkout ${github.getBaseSha()}`);
  // base の Graph を生成
  const { graph: fullBaseGraph } = createGraph({
    tsconfig: getTsconfigPath(),
    dir: path.resolve(getTsconfigRoot()),
  });
  log('fullBaseGraph.nodes.length:', fullBaseGraph.nodes.length);
  log('fullBaseGraph.relations.length:', fullBaseGraph.relations.length);
  return { fullHeadGraph, fullBaseGraph, meta };
}
