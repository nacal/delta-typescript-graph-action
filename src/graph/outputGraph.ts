import mermaidify from '@ysk8hori/typescript-graph/dist/src/mermaidify';
import { Graph, Meta } from '@ysk8hori/typescript-graph/dist/src/models';
import { getMaxSize, getOrientation, isInDetails } from '../utils/config';
import mergeGraphsWithDifferences from './mergeGraphsWithDifferences';
import github from '../github';
import { info } from '../utils/log';

type FileInfoList = {
  filename: string;
  previous_filename: string | undefined;
}[];

export function outputGraph(
  fullBaseGraph: Graph,
  fullHeadGraph: Graph,
  meta: Meta,
  files: {
    created: FileInfoList;
    deleted: FileInfoList;
    modified: FileInfoList;
    renamed: FileInfoList;
  },
) {
  const { graph, tsgCommand } = mergeGraphsWithDifferences(
    fullBaseGraph,
    fullHeadGraph,
    files.created.map(({ filename }) => filename),
    files.deleted.map(({ filename }) => filename),
    files.modified.map(({ filename }) => filename),
    files.renamed,
  );

  if (graph.nodes.length === 0) {
    // グラフが空の場合は表示しない
    info('The graph is empty.');
    return;
  }

  if (graph.nodes.length > getMaxSize()) {
    // グラフが大きすぎる場合は表示しない
    github.commentToPR(`
${github.getCommentTitle()}

${outputIfInDetails(`
<details>
<summary>mermaid</summary>
`)}

> 表示ノード数が多いため、グラフを表示しません。
> 表示ノード数の上限を変更したい場合はアクションのパラメータ \`max-size\` を設定してください。
>
> 本PRでの表示ノード数: ${graph.nodes.length}
> 最大表示ノード数: ${getMaxSize()}

\`\`\`bash
${tsgCommand}
\`\`\`

${outputIfInDetails('</details>')}
`);
    return;
  }

  const mermaidLines: string[] = [];
  mermaidify((arg: string) => mermaidLines.push(arg), graph, {
    rootDir: meta.rootDir,
    ...getOrientation(),
  });

  github.commentToPR(`
${github.getCommentTitle()}

${outputIfInDetails(`
<details>
<summary>mermaid</summary>
`)}

\`\`\`bash
${tsgCommand}
\`\`\`

\`\`\`mermaid
${mermaidLines.join('')}
\`\`\`

${outputIfInDetails('</details>')}
`);
}

/** isMermaidInDetails() の結果が true ならば与えられた文字列を返し、そうでなければ空文字を返す関数。 */
function outputIfInDetails(str: string): string {
  return isInDetails() ? str : '';
}
