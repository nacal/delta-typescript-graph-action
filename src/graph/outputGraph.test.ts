import { Graph, Node } from '@ysk8hori/typescript-graph/dist/src/models';
import { outputGraph } from './outputGraph';
import github from '../github';

github.commentToPR = jest.fn();
github.deleteComment = jest.fn();
github.getCommentTitle = jest.fn();
(github.getCommentTitle as jest.Mock).mockImplementation(
  () => `## Delta TypeScript Graph<!--test-workflow.yml-->`,
);

const a: Node = {
  path: 'src/A.tsx',
  name: 'A',
  changeStatus: 'not_modified',
};
const b: Node = {
  path: 'src/B.tsx',
  name: 'B',
  changeStatus: 'not_modified',
};
const c: Node = {
  path: 'src/C.tsx',
  name: 'C',
  changeStatus: 'not_modified',
};

const d: Node = {
  path: 'src/1/D.tsx',
  name: 'D',
  changeStatus: 'not_modified',
};

const e: Node = {
  path: 'src/1/E.tsx',
  name: 'E',
  changeStatus: 'not_modified',
};

test('出力可能なグラフがない場合は何も出力しない', () => {
  const graph = {
    nodes: [],
    relations: [],
  };
  const meta = {
    rootDir: '',
  };
  outputGraph(graph, graph, meta, {
    created: [],
    deleted: [],
    modified: [],
    renamed: [],
  });
  expect(github.commentToPR).not.toHaveBeenCalled();
});

test('追加や依存の削除がある場合', () => {
  const graphA: Graph = {
    nodes: [a, c, d, e],
    relations: [
      {
        from: a,
        to: c,
        kind: 'depends_on',
        fullText: '',
        changeStatus: 'not_modified',
      },
      {
        from: a,
        to: d,
        kind: 'depends_on',
        fullText: '',
        changeStatus: 'not_modified',
      },
      {
        from: d,
        to: e,
        kind: 'depends_on',
        fullText: '',
        changeStatus: 'not_modified',
      },
    ],
  };
  const graphB: Graph = {
    nodes: [a, b, c, d, e],
    relations: [
      {
        from: a,
        to: b,
        kind: 'depends_on',
        fullText: '',
        changeStatus: 'not_modified',
      },
      {
        from: a,
        to: d,
        kind: 'depends_on',
        fullText: '',
        changeStatus: 'not_modified',
      },
      {
        from: d,
        to: e,
        kind: 'depends_on',
        fullText: '',
        changeStatus: 'not_modified',
      },
    ],
  };
  const meta = {
    rootDir: '',
  };
  outputGraph(graphA, graphB, meta, {
    created: [{ filename: b.path, previous_filename: undefined }],
    deleted: [],
    modified: [{ filename: a.path, previous_filename: undefined }],
    renamed: [],
  });
  expect((github.commentToPR as jest.Mock).mock.calls[0][0]).toMatchSnapshot();
});
