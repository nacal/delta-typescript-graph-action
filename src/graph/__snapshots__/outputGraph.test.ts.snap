// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`追加や依存の削除がある場合 1`] = `
"
## Delta TypeScript Graph<!--test-workflow.yml-->



\`\`\`bash
tsg --include src/B.tsx src/A.tsx --highlight src/B.tsx src/A.tsx --exclude node_modules --abstraction src/1
\`\`\`

\`\`\`mermaid
flowchart
    classDef created fill:cyan,stroke:#999,color:black
    classDef modified fill:yellow,stroke:#999,color:black
    subgraph src["src"]
        src/A.tsx["A"]:::modified
        src/B.tsx["B"]:::created
        src/1["/1"]:::dir
        src/C.tsx["C"]
    end
    src/A.tsx-->src/B.tsx
    src/A.tsx-->src/1
    src/A.tsx-.->src/C.tsx

\`\`\`


"
`;
